tests/unit_tests/io/io_json_payload.cpp

```diff
@@ -0,0 +1,374 @@
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#include <gtest/gtest.h>
+
+#include "detray/io/json_io.hpp"
+
+TEST(io, json_algebra_payload) {
+
+    detray::transform_payload p;
+    p.tr = {100., 200., 300.};
+    p.rot = {1., 0., 0., 0., 1., 0., 0., 0., 1};
+
+    nlohmann::json j;
+    j["transform"] = p;
+
+    detray::transform_payload pt = j["transform"];
+
+    EXPECT_EQ(p.tr, pt.tr);
+    EXPECT_EQ(p.rot, pt.rot);
+}
+
+TEST(io, json_axis_payload) {
+
+    detray::axis_payload ea;
+    ea.type = detray::axis_payload::axis_type::equidistant;
+    ea.bracket = detray::axis_payload::axis_bracket::closed;
+    ea.lookup = detray::axis_payload::axis_lookup::phi;
+    ea.borders = {-M_PI, M_PI};
+    ea.bins = 10u;
+
+    nlohmann::json je;
+    je["axis"] = ea;
+
+    detray::axis_payload pea = je["axis"];
+
+    EXPECT_EQ(ea.type, pea.type);
+    EXPECT_EQ(ea.bracket, pea.bracket);
+    EXPECT_EQ(ea.borders, pea.borders);
+
+    EXPECT_EQ(ea.bins, pea.bins);
+
+    detray::axis_payload va;
+    va.type = detray::axis_payload::axis_type::variable;
+    va.bracket = detray::axis_payload::axis_bracket::bound;
+    va.lookup = detray::axis_payload::axis_lookup::r;
+    va.borders = {0, 1, 4, 5, 8, 10};
+    va.bins = va.borders.size() - 1;
+
+    nlohmann::json jv;
+    jv["axis"] = va;
+
+    detray::axis_payload pva = jv["axis"];
+
+    EXPECT_EQ(va.type, pva.type);
+    EXPECT_EQ(va.bracket, pva.bracket);
+    EXPECT_EQ(va.lookup, pva.lookup);
+    EXPECT_EQ(va.borders, pva.borders);
+    EXPECT_EQ(va.bins, pva.bins);
+}
+
+TEST(io, json_grid_payload) {
+
+    std::vector<std::vector<unsigned int>> entries = {{0, 1}, {0, 2}, {1, 1},
+                                                      {1, 2}, {2, 1}, {2, 2}};
+
+    detray::axis_payload a0{detray::axis_payload::axis_type::equidistant,
+                            detray::axis_payload::axis_bracket::closed,
+                            detray::axis_payload::axis_lookup::phi,
+                            std::vector<detray::real_io>{-M_PI, M_PI}, 3u};
+
+    detray::axis_payload a1{detray::axis_payload::axis_type::equidistant,
+                            detray::axis_payload::axis_bracket::bound,
+                            detray::axis_payload::axis_lookup::r,
+                            std::vector<detray::real_io>{0, 2u}, 2u};
+
+    detray::grid_payload g;
+    g.axes = {a0, a1};
+    g.entries = entries;
+
+    nlohmann::json j;
+    j["grid"] = g;
+
+    detray::grid_payload pg = j["grid"];
+
+    EXPECT_EQ(g.axes.size(), pg.axes.size());
+    EXPECT_EQ(g.entries, pg.entries);
+}
+
+TEST(io, single_object_payload) {
+    detray::single_object_payload so;
+    so.link = 3u;
+
+    nlohmann::json j;
+    j["single_link"] = so;
+
+    detray::single_object_payload pso = j["single_link"];
+
+    EXPECT_EQ(so.link, pso.link);
+}
+
+TEST(io, grid_objects_payload) {
+    detray::grid_objects_payload go;
+
+    std::vector<std::vector<unsigned int>> entries = {{0, 1}, {0, 2}, {1, 1},
```

Comment:
Hm, a question in general: should we repeat the grid, transform tests etc. for every higher level component?

Reply:
Let's see ...